src/lua/msgpack.c

```diff
@@ -809,6 +815,80 @@ luamp_object_iterator(struct lua_State *L)
 	return 1;
 }
 
+/**
+ * Decodes the MsgPack data stored in @a obj and saves a reference to it on the
+ * Lua stack.
+ */
+static void
+luamp_object_get_decode(struct lua_State *L, struct luamp_object *obj)
+{
+	if (obj->decoded_ref != LUA_NOREF)
+		return;
+	const char *data = obj->data;
+	luamp_decode(L, obj->cfg, &data);
+	assert(data == obj->data_end);
+	obj->decoded_ref = luaL_ref(L, LUA_REGISTRYINDEX);
+}
+
+/**
+ * Takes a `msgpack.object` and an indexation key as the arguments, indexes
+ * the MsgPack stored in the `msgpack.object` and pushes the result to Lua stack
+ * or throws an error.
+ */
+static int
+luamp_object_get(struct lua_State *L)
+{
+	struct luamp_object *obj = luamp_check_object(L, 1);
+	enum mp_type type = mp_typeof(*obj->data);
+	if (type != MP_MAP && type != MP_ARRAY)
+		return 0;
+	luamp_object_get_decode(L, obj);
+	/* Pushes the decoded MsgPack data on top of the stack. */
+	lua_rawgeti(L, LUA_REGISTRYINDEX, obj->decoded_ref);
+	/* Pushes the indexing key on top of the stack. */
+	lua_pushvalue(L, -2);
+	/* Indexes the decoded MsgPack data and pops the key. */
+	lua_rawget(L, -2);
+	/* Replaces the decoded MsgPack data with the indexing result. */
+	lua_replace(L, -2);
+	return 1;
+}
+
+/**
+ * Takes a `msgpack.object` and an indexation key as the arguments: if the key
+ * is of string type, first tries to match it with `msgpack.object` methods,
+ * and, in case the match occurs, pushes the matched method to Lua stack —
+ * otherwise, delegates indexation to `msgpack.object:get`.
+ */
+static int
+luamp_object_index(struct lua_State *L)
+{
+	if (lua_type(L, 2) != LUA_TSTRING)
+		return luamp_object_get(L);
+
+	lua_getmetatable(L, 1);
+	lua_pushvalue(L, 2);
```

Comment:
This may crash if misused:

```
tarantool> require('msgpack').object(123).__index(123, 'foo')
tarantool: ./third_party/luajit/src/lj_api.c:840: lua_rawget: Assertion `(((t)->it) == (~11u))' failed.
Aborted
```

Please fix.

Reply:
Fixed: check that the first argument is `msgpack.object` in `__index` and `__autocomplete`.

Added test cases for this.